packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -305,271 +297,180 @@ class FadeInImage extends StatefulWidget {
-  });
-
-  final _FadeInImageState state;
-  final _ImageProviderResolverListener listener;
-
-  FadeInImage get widget => state.widget;
+  /// This description will be used both while the [placeholder] is shown and
+  /// once the image has loaded.
+  final String imageSemanticLabel;
 
-  ImageStream _imageStream;
-  ImageInfo _imageInfo;
+  Image _image({
+    @required ImageProvider image,
+    ImageFrameBuilder frameBuilder,
+  }) {
+    assert(image != null);
+    return Image(
+      image: image,
+      frameBuilder: frameBuilder,
+      width: width,
+      height: height,
+      fit: fit,
+      alignment: alignment,
+      repeat: repeat,
+      matchTextDirection: matchTextDirection,
+      gaplessPlayback: true,
+      excludeFromSemantics: true,
+    );
+  }
 
-  void resolve(ImageProvider provider) {
-    final ImageStream oldImageStream = _imageStream;
-    _imageStream = provider.resolve(createLocalImageConfiguration(
-      state.context,
-      size: widget.width != null && widget.height != null ? Size(widget.width, widget.height) : null,
-    ));
-    assert(_imageStream != null);
+  @override
+  Widget build(BuildContext context) {
+    Widget result = _image(
+      image: image,
+      frameBuilder: (BuildContext context, Widget child, int frame, bool wasSynchronouslyLoaded) {
+        if (wasSynchronouslyLoaded)
+          return child;
+        return _AnimatedFadeOutFadeIn(
+          image: child,
+          placeholder: _image(image: placeholder),
+          isTargetLoaded: frame != null,
+          fadeInDuration: fadeInDuration,
+          fadeOutDuration: fadeOutDuration,
+          fadeInCurve: fadeInCurve,
+          fadeOutCurve: fadeOutCurve,
+        );
+      },
+    );
 
-    if (_imageStream.key != oldImageStream?.key) {
-      final ImageStreamListener listener = ImageStreamListener(_handleImageChanged);
-      oldImageStream?.removeListener(listener);
-      _imageStream.addListener(listener);
+    if (!excludeFromSemantics) {
+      result = Semantics(
+        container: imageSemanticLabel != null,
+        image: true,
+        label: imageSemanticLabel ?? '',
+        child: result,
+      );
     }
-  }
-
-  void _handleImageChanged(ImageInfo imageInfo, bool synchronousCall) {
-    _imageInfo = imageInfo;
-    listener();
-  }
 
-  void stopListening() {
-    _imageStream?.removeListener(ImageStreamListener(_handleImageChanged));
+    return result;
   }
 }
 
-class _FadeInImageState extends State<FadeInImage> with TickerProviderStateMixin {
-  _ImageProviderResolver _imageResolver;
-  _ImageProviderResolver _placeholderResolver;
-
-  AnimationController _controller;
-  Animation<double> _animation;
-
-  FadeInImagePhase _phase = FadeInImagePhase.start;
-  FadeInImagePhase get phase => _phase;
+class _AnimatedFadeOutFadeIn extends ImplicitlyAnimatedWidget {
+  const _AnimatedFadeOutFadeIn({
+    Key key,
+    @required this.image,
+    @required this.placeholder,
+    @required this.isTargetLoaded,
```

Comment:
rename `image` to target or rename this to `isImageLoaded` to show the relationship between the two?

Reply:
Done